Add global uses limit to promotions

Pebblo opened the pull request:
See #26 

## Problem this Pull Request solves
This adds a 'Global number of Uses' field to the promotions create/edit screen, the logic for 'global uses' is already set within the add-on.

## How has this been tested
Create a promotion code and set the global uses to 1.
(Set the other uses to something higher, like 11 and it will basically be ignored with the global value set)

Use the promotion code to the uses limit and then confirm that it cant not be used again on any event it applies to.

The promotions does **not** need to be global for this global limit to apply.

Thoughts on maybe changing the name? I get the feelinh using 'Global' here might imply this needs to be global.

## Checklist

* [ ] I have read the documentation relating to systems affected by this pull request, see https://github.com/eventespresso/event-espresso-core/tree/master/docs
* [ ] User input is adequately validated and sanitized
* [ ] all publicly displayed strings are internationalized (usually using `esc_html__()`, see https://codex.wordpress.org/I18n_for_WordPress_Developers)
* [ ] My code is tested.
* [ ] My code follows the Event Espresso code style.
* [ ] My code has proper inline documentation.

tn3rb commented:
>  I get the feeling using 'Global' here might imply this needs to be global.

IIRC, setting a promo as global assigned it to ALL events, whereas setting the global uses would limit the total number of uses across however many events the promo was assigned to, which does not have to be ALL events (ie: a global promo)

examples:

```
isGlobal === true
globalUses === 10
```

can be used a total of 10 times on ANY event


```
isGlobal === false
globalUses === 10
```

can be used a total of 10 times, but just for the events it has been assigned to

tn3rb commented on admin/promotions/templates/promotion_details_form.template.php:
```diff
@@ -71,7 +71,7 @@
     <tr>
         <th scope="row">
-            <label for="PRO_uses"><?php _e('Apply Promo to ALL Scope Items', 'event_espresso'); ?></label>
+            <label for="PRO_global"><?php _e('Apply Promo to ALL Scope Items', 'event_espresso'); ?></label>
```
good catch 👍

tn3rb commented on admin/promotions/templates/promotion_details_form.template.php:
```diff
@@ -80,7 +80,7 @@
     <tr>
         <th scope="row">
-            <label for="PRO_uses"><?php _e('Promo Is Exclusive', 'event_espresso'); ?></label>
+            <label for="PRO_exclusive"><?php _e('Promo Is Exclusive', 'event_espresso'); ?></label>
```
another good catch 👍

tn3rb commented on admin/promotions/templates/promotion_details_form.template.php:
```diff
@@ -96,6 +96,15 @@
+            <input type="text" class="regular-text ee-numeric" id="PRO_global_uses" name="PRO_global_uses" value="<?php echo $promotion_global_uses; ?>">
+            <p class="description"><?php _e('This determines how many times this promotion code can be applied across all scopes, this value overrides the scope uses field if set at a lower value - leave blank for no limit', 'event_espresso'); ?></p>
+        </td>
+    </tr>
```
let's move this up so that it appears immediately after the `PRO_global` input so that those two settings are grouped together.

As well, can you plz change all of the text strings in this template to use `esc_html_e()` instead of just `_e()`?

Pebblo replied:
Oh yeah! Tbh, I'd completely forgot about doing that.

WRT moving the above, I've done it but I think this may further point towards users thinking this is a limit for global promo codes, will add a comment below.

Pebblo commented:
> IIRC, setting a promo as global assigned it to ALL events, whereas setting the global uses would limit the total number of uses across however many events the promo was assigned to, which does not have to be ALL events (ie: a global promo)

Yeah that's correct, this new value is the 'one to rule them all', this limit allows the promotion code to be used that many times, regardless if its global or not.

However, I think the text 'Global number of Uses' might confuse users into thinking this limit is for Global promo codes only and if they want to use that value must set the promotion to global.

So I'm wondering if we should try to label it differently?

tn3rb commented:
> However, I think the text 'Global number of Uses' might confuse users into thinking this limit is for Global promo codes only and if they want to use that value must set the promotion to global.
> 
> So I'm wondering if we should try to label it differently?

hmmmmm ya... very good point

how about simply 'Total number of Uses' ?

joshfeck commented:
I'd be OK with "Total Number of Uses" as @tn3rb mentioned. Or another idea: "Limit uses to".

Then I think we could also help clarify the difference with the existing "Number of Uses" field by changing its label to: "Number of Uses Per Scope Item"

Pebblo commented:
Updated the text to the above and also updated the 'normal' uses filed text to show it applies per scope item...

Seeing how that was a (translation) breaking change and adding a field into the mix will already mess with peoples muscle memory, I re-ordered this a little more so it flows a little better imo.

joshfeck commented:
![Screen Shot 2020-03-09 at 4 48 17 PM](https://user-images.githubusercontent.com/891156/76331747-e3506e80-62c5-11ea-9219-84811c34436b.png)

joshfeck commented:
![2020-03-10 11 35 07](https://user-images.githubusercontent.com/891156/76331905-15fa6700-62c6-11ea-8e49-db5e3d3f9c41.png)

joshfeck commented:
Not related to the changes in the PR: There might be an edge case issue for sites using MER (where two different events in the cart will each get a discount even if 1 global use is set/left). That's something within the logic that was already in place, now that we have a UI to set the Global use option that logic may need to be adjusted to avoid this.

Pebblo commented:
First commit checks the available global uses before each promotion line item is added (the uses value is updated in `add_promotion_line_item()` )

The second commit updates the 'redeemed' column in the promotions list table to also use the global uses setting.

Pebblo commented:
This latest change prevented me from applying an older promotion code when testing something else so pulling back for now.

sethshoultes commented:
This seems to be working okay. However, I noticed that if you have two different events, then set the Number of Uses Per Scope item to 2 and set Total Number of Uses to 4, it shows that the promo code is Unavailable when there's only one use left:
![image](https://user-images.githubusercontent.com/235315/77358936-6d5df580-6d10-11ea-98c6-9f9e6d7ea54d.png)


@Pebblo also, can you please capitalize the field labels? 
![‹ Event Espresso Testing — WordPress 2020-03-23 13-56-34(1)](https://user-images.githubusercontent.com/235315/77358265-24597180-6d0f-11ea-83bd-e4536faffd55.jpg)

@lorenzocaum and @garthkoyle adding you to this PR so you are aware of this update.

joshfeck commented:
> I noticed that if you have two different events, then set the Number of Uses Per Scope item to 2 and set Total Number of Uses to 4, it shows that the promo code is Unavailable when there's only one use left

That looks like a separate issue (also in Master) where the code that gets the status doesn't get a count of available uses. We can fix that in a separate PR.

joshfeck commented:
> also, can you please capitalize the field labels?

This is done.

Pebblo commented:
Hat tip to @joshfeck for the above fix as it's his, I just pushed it.

The above change should make the status of the promotion take usage for multiple scope items into account.

Pushed to this branch as it is only used in the admin and can be easily tested whilst testing the changes here.

@sethshoultes if the above fixes your promotion status I'd say this is ready to go.

joshfeck commented:
I also confirmed the other changes helped avoid the issue when two events in the same cart could both get a promotion applied when only one global use was left.

Pebblo commented:
Merging this in as I think it's been tested enough.
